### PR TITLE
Improve access of typed versions of components

### PR DESCRIPTION
**Pull request by dominiklohmann**
###  :notebook_with_decorative_cover: Description

This commit
- replaces all usage of the untyped version of get_node_components with
  the typed version,
- removes the untpyed version of get_node_components entirely,
- introduces a new typed find member function to the component registry,
  and
- replaces calls to find_by_label with calls to the new typed find where
  possible.

###  :memo: Checklist

- [x] All user-facing changes have changelog entries.
- [x] The changes are reflected on [docs.tenzir.com/vast](https://docs.tenzir.com/vast), if necessary.
- [x] The PR description contains instructions for the reviewer, if necessary.

### :dart: Review Instructions

File-by-file.

### CODE REVIEW

File: libvast/vast/detail/actor_cast_wrapper.hpp
Comment (mavam) [inline libvast/vast/detail/actor_cast_wrapper.hpp:24]: Can you briefly explain in a comment what this does and/or give an example?

File: libvast/src/system/node.cpp
Comment (mavam) [inline libvast/src/system/node.cpp:572]: It feels like you could if-constexpr on the size of the `Ts` and then either return a single value or a tuple, as before. The singleton-wrapping feels a bit awkward.

File: libvast/src/system/node.cpp
Comment (mavam) [inline libvast/src/system/node.cpp:521]: Nope 😉

File: libvast/vast/system/node_control.hpp
Comment (mavam) [inline libvast/vast/system/node_control.hpp:44]: I wish we could have a better name for this function, but I can't come up with one...

File: libvast/vast/system/component_registry.hpp
Comment (mavam) [inline libvast/vast/system/component_registry.hpp:93]: This computation requires some assertions, e.g., `starts_with(str, "vast::system::")`.

Unless the registration with labe also works as dual to this transformation, I don't feel comfortable with this method. @tobim?

**Review by mavam — COMMENTED** [review]
Nice changes from a user-perspective.

The type-name-to-label mapping is a bit fishy and needs to made explicit. The inverse must also exist at the registration site for this to be safe. I would just write a trait that gives you the label for a given component actor type.

File: libvast/vast/system/component_registry.hpp
Comment (dominiklohmann) [reply libvast/vast/system/component_registry.hpp:93]: Note that this transformation code was written by @tobim originally. Let's talk about this tomorrow briefly. I think it's fine as is for now and we can change it up when making the node itself a typed actor.

Comment (mavam) [reply libvast/vast/system/component_registry.hpp:93]: Good to know. I think it's actually not complicated to switch from

```cpp
const char* singletons[]
    = {"accountant", "archive", "eraser",       "filesystem",
       "importer",   "index",   "type-registry"};
```

to a type list in the node. My main worry is that these go out of sync and then we're in trouble.

I agree it's best to chat about this briefly.

File: libvast/src/system/node.cpp
Comment (dominiklohmann) [reply libvast/src/system/node.cpp:572]: It's awkward here, but doing one-offs is just extra maintenance and makes working with it in generic contexts non-trivial. I'd rather always keep it a tuple.

File: libvast/vast/detail/actor_cast_wrapper.hpp
Comment (dominiklohmann) [reply libvast/vast/detail/actor_cast_wrapper.hpp:24]:
```cpp
actor_cast_wrapper{}
```

is essentially this from C++20:

```cpp
[]<typename Out>(auto&& in) {
  return caf::actor_cast<Out>(
    std::forward<decltype(in)>(in));
}
```

I'll add that as a comment with a note to replace once switching to C++20.

File: libvast/src/system/node.cpp
Comment (dominiklohmann) [reply libvast/src/system/node.cpp:521]: I suppose this is what happens if you set `,` as leader. 🙈

File: libvast/vast/system/component_registry.hpp
Comment (tobim) [reply libvast/vast/system/component_registry.hpp:93]: I agree with the general sentiment that these should be generated from the same expressions or checked statically. 

> My main worry is that these go out of sync and then we're in trouble.

When one changes one but not the other during development we'll see some failing unit or integration tests. That won't give a good indication of what is wrong, and makes the dev waste time on debugging.

Comment (tobim) [reply libvast/vast/system/component_registry.hpp:93]: @dominiklohmann is it feasible to generate `singletons` from the type ids and centralize (or drop?) the normalization?

File: libvast/vast/detail/actor_cast_wrapper.hpp
Comment (tobim) [reply libvast/vast/detail/actor_cast_wrapper.hpp:24]: It's quite annoying that we have to add those one-offs because of a lack of orthogonality in the language.

Comment (dominiklohmann) [reply libvast/vast/detail/actor_cast_wrapper.hpp:24]: Even more so because these cannot be defined anonymously within functions. 😤

File: libvast/vast/system/component_registry.hpp
Comment (dominiklohmann) [reply libvast/vast/system/component_registry.hpp:93]: Let's do this only after we converted the node to a typed actor interface. It's almost impossible to catch all possible issues with this currently.

Comment (dominiklohmann) [reply libvast/vast/system/component_registry.hpp:93]: I added a comment above `singletons` that explains what we should do here in the future.

File: libvast/vast/system/node_control.hpp
Comment (dominiklohmann) [reply libvast/vast/system/node_control.hpp:44]: Let's not change that within this PR.

File: libvast/src/system/node.cpp
Comment (dominiklohmann) [reply libvast/src/system/node.cpp:586]: @tobim seems to agree with me here—marking this as resolved.

File: libvast/vast/system/component_registry.hpp
Comment (dominiklohmann) [reply libvast/vast/system/component_registry.hpp:93]: @tobim does my comment suffice? Can you re-review?

Comment (mavam) [reply libvast/vast/system/component_registry.hpp:93]: > Let's do this only after we converted the node to a typed actor interface.

Alright, fine with me. Can you take care of creating a corresponding story?

Comment (dominiklohmann) [reply libvast/vast/system/component_registry.hpp:93]: Will do before merging this.

Comment (dominiklohmann) [reply libvast/vast/system/component_registry.hpp:93]: Two stories exist now as a follow-up. @tobim / @mavam mind approving (unless you want anything else changed)?

**Review by tobim — APPROVED** [review]
Reducing the risk for error is always welcome.

**Pull request merged**
